### ISSUE TITLE
How to handle `result: null`?

### ISSUE THREAD

**Issue by dpc** [apoelstra/rust-jsonrpc#15]
For bitcoind, if I call `invalidateblock`, the `result` is `null`:

```
curl --user user  --data-binary '{"jsonrpc": "2.0", "id":"curltest", "method": "invalidateblock", "params": ["0d99e36fe7b3a91b01dc393da7c5e900c17a28390a4ba60820eaa51af60a029e"] }' -H 'content-type: text/plain;' http://127.0.0.1:45929/
Enter host password for user 'user':
{"result":null,"error":null,"id":"curltest"}
```
Make no mistake. The operation was successful.

However it turns into:

```
Response: Err(NoErrorOrResult

Response format is invalid)
```

when calling `into_result`. 

https://www.jsonrpc.org/specification#response_object states:
```
result
    This member is REQUIRED on success.
    This member MUST NOT exist if there was an error invoking the method.
```

I'm not sure if `result: null` is conforming to the standard  or not, but I side with "yes". There is no result to return, so it's `null`.

Does it mean that `into_resul()` is handling this case incorectly, or should I be just not calling `into_result()`? It seems to me that it would be better if it wasn't neccessary to call different methods, depending on the expected result (or lack of it).

**Comment by dpc** [comment]
Ping @jeandudey as bitcoin-rpc library is affected (I'll try to submit a PR), as I'm working on it.

**Comment by apoelstra** [comment]
Thanks Dawid!,

I agree that Core's behaviour here is perfectly reasonable. Unfortunately serde(-json) makes it hard to distinguish nulls from missing fields.

The simplest solution may be to just remove this error path and accept that nonconforming implementations that fail to provide either field will get away with it.



On Sun, Oct 28, 2018 at 11:42:45PM -0700, Dawid Ciężarkiewicz wrote:
> Ping @jeandudey as bitcoin-rpc library is affected (I'll try to submit a PR), as I'm working on it.
> 
> -- 
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly or view it on GitHub:
> https://github.com/apoelstra/rust-jsonrpc/issues/15#issuecomment-433803518

-- 
Andrew Poelstra
Research Director, Mathematics Department, Blockstream
Email: apoelstra at wpsoftware.net
Web:   https://www.wpsoftware.net/andrew

"Make it stop, my love; we were wrong to try
 Never saw what we could unravel in traveling light
 Nor how the trip debrides like a stack of slides
 All we saw was that time is taller than space is wide"
       --Joanna Newsom

**Comment by dpc** [comment]
I agree. The spec doesn't say that receive has to validate if the field is really there, so it is a reasonable solution.

**Comment by dpc** [comment]
Now that I look at the code - maybe there is a way to it, only if `T` in `pub fn into_result<T: serde::de::DeserializeOwned>(self) -> Result<T, Error> {` is `()`.

**Comment by stevenroose** [comment]
Can we have a release with this? I'm hitting this too.

**Comment by apoelstra** [comment]
Yep. This was fixed in #16 I believe and released in #25, a couple days ago.
